File: Thawbuk-store/lib/data/models/product_model.dart
import 'package:json_annotation/json_annotation.dart';
import '../../domain/entities/product_entity.dart';

part 'product_model.g.dart';

@JsonSerializable(explicitToJson: true)
class AgeRangeModel extends AgeRangeEntity {
  const AgeRangeModel({
    required super.minAge,
    required super.maxAge,
  });

  factory AgeRangeModel.fromJson(Map<String, dynamic> json) =>
      _$AgeRangeModelFromJson(json);

  Map<String, dynamic> toJson() => _$AgeRangeModelToJson(this);
}

@JsonSerializable(explicitToJson: true)
class ProductModel extends ProductEntity {
  @override
  @JsonKey(name: 'ageRange')
  final AgeRangeModel? ageRange;

  const ProductModel({
    required this.id,
    required String name,
    String? nameAr,
    String? description,
    String? descriptionAr,
    required double price,
    required String categoryId,
    String? createdBy,
    required List<String> images,
    List<String>? sizes,
    List<String>? colors,
    required int stock,
    DateTime? createdAt,
    DateTime? updatedAt,
    double? rating,
    int? reviewsCount,
    String? brand,
    Map<String, int>? ageRange,
    int? favoritesCount,
    int? viewsCount,
    bool? isActive,
  }) : super(
          id: id,
          name: name,
          nameAr: nameAr,
          description: description,
          descriptionAr: descriptionAr,
          price: price,
          categoryId: categoryId,
          createdBy: createdBy,
          images: images,
          sizes: sizes,
          colors: colors,
          stock: stock,
          createdAt: createdAt,
          updatedAt: updatedAt,
          rating: rating,
          reviewsCount: reviewsCount,
          brand: brand,
          ageRange: ageRange,
          favoritesCount: favoritesCount,
          viewsCount: viewsCount,
          isActive: isActive,
        );
>>>>>>> main

  factory ProductModel.fromJson(Map<String, dynamic> json) =>
      _$ProductModelFromJson(json);

  Map<String, dynamic> toJson() => _$ProductModelToJson(this);
  ProductEntity toEntity() {
    return ProductEntity(
      id: id,
      name: name,
      nameAr: nameAr,
      description: description,
      descriptionAr: descriptionAr,
      price: price,
      categoryId: categoryId,
      createdBy: createdBy,
      images: images,
      sizes: sizes,
      colors: colors,
      stock: stock,
      createdAt: createdAt,
      updatedAt: updatedAt,
      rating: rating,
      reviewsCount: reviewsCount,
      brand: brand,
      ageRange: ageRange,
      favoritesCount: favoritesCount,
      viewsCount: viewsCount,
      isActive: isActive,
    );
  }
}

File: src/infrastructure/database/mongodb/models/product.model.ts
import mongoose, { Schema, Model, Document } from "mongoose";
import { IProduct } from "../../../../domain/entity/product";

type ProductDocument = IProduct & Document;

const productSchema = new Schema<ProductDocument>({
    _id: { type: String, default: () => crypto.randomUUID() },
    name: { type: String, required: true },
    price: { type: Number, required: true },
    description: { type: String },
    images: [{ type: String, match: [/^https?:\/\/.+/] }],
    sizes: [{ type: String }],
    colors: [{ type: String }],
    stock: { type: Number, default: 0 },
    brand: { type: String },
    ageRange: {
        minAge: { type: Number },
        maxAge: { type: Number },
    },
    categoryId: { type: String, ref: 'Category', required: true },
    createdBy: {
        type: String,
        ref: 'User',
        required: true,
    },
}, {
    timestamps: true,
});

productSchema.index({ title: 'text' });
productSchema.index({ categoryId: 1 });
productSchema.index({ createdAt: -1 });


export const ProductModel: Model<ProductDocument> = mongoose.model<ProductDocument>('Product', productSchema);